pilot/pkg/model/extensions.go

```diff
@@ -0,0 +1,121 @@
+	envoy_extensions_filters_http_wasm_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/wasm/v3"
+	envoy_extensions_wasm_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/wasm/v3"
+	"github.com/gogo/protobuf/types"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	extensions "istio.io/api/extensions/v1alpha1"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/util/gogo"
+)
+
+const (
+	defaultRuntime = "envoy.wasm.runtime.v8"
+)
+
+type WasmPluginWrapper struct {
+	extensions.WasmPlugin
+
+	Name      string
+	Namespace string
+
+	ExtensionConfiguration *envoy_extensions_filters_http_wasm_v3.Wasm
+}
+
+func convertToWasmPluginWrapper(plugin *config.Config) *WasmPluginWrapper {
+	if wasmPlugin, ok := plugin.Spec.(*extensions.WasmPlugin); ok {
+		cfg := &any.Any{}
+		if wasmPlugin.PluginConfig != nil && len(wasmPlugin.PluginConfig.Fields) > 0 {
+			cfgJSON, _ := protojson.Marshal(proto.MessageV2(wasmPlugin.PluginConfig))
+			cfg = gogo.MessageToAny(&types.StringValue{
+				Value: string(cfgJSON),
+			})
+		}
+		var datasource *envoy_config_core_v3.AsyncDataSource
+		var sha256 string
+		u, err := url.Parse(wasmPlugin.Url)
+		if err != nil {
+			return nil
+		}
+		if wasmPlugin.XSha256 == nil {
+			if u.Scheme == "http" || u.Scheme == "https" {
+				// SHA256 is required for .wasm deployments fetched from HTTP/HTTPS URLs
+				return nil
+			}
+			// field is required, so we're setting a string for unmarshaling to not fail
+			// on the agent side. this will never reach envoy
+			sha256 = "nil"
+		} else {
+			sha256 = wasmPlugin.GetSha256()
+		}
+		if u.Scheme == "file" {
+			datasource = &envoy_config_core_v3.AsyncDataSource{
+				Specifier: &envoy_config_core_v3.AsyncDataSource_Local{
+					Local: &envoy_config_core_v3.DataSource{
+						Specifier: &envoy_config_core_v3.DataSource_Filename{
+							// remove 'file://' prefix
+							Filename: wasmPlugin.Url[7:],
+						},
+					},
+				},
+			}
+		} else {
+			datasource = &envoy_config_core_v3.AsyncDataSource{
+				Specifier: &envoy_config_core_v3.AsyncDataSource_Remote{
+					Remote: &envoy_config_core_v3.RemoteDataSource{
+						HttpUri: &envoy_config_core_v3.HttpUri{
+							Uri:     wasmPlugin.Url,
+							Timeout: durationpb.New(30 * time.Second),
+							HttpUpstreamType: &envoy_config_core_v3.HttpUri_Cluster{
+								// this will be fetched by the agent anyway, so no need for a cluster
+								Cluster: "_",
+							},
+						},
+						Sha256: sha256,
+					},
+				},
+			}
+		}
+		return &WasmPluginWrapper{
+			Name:       plugin.Name,
+			Namespace:  plugin.Namespace,
+			WasmPlugin: *wasmPlugin,
+			ExtensionConfiguration: &envoy_extensions_filters_http_wasm_v3.Wasm{
+				Config: &envoy_extensions_wasm_v3.PluginConfig{
+					Name:          plugin.Name,
+					RootId:        wasmPlugin.PluginName,
```

Comment:
Why do we set root id here? This will need to be compiled into the Wasm code. We should leave it unset.

Reply:
iiuc, root id should also be set here, not only in a Wasm binary. this is only used by C++ and AssemblyScript SD, and rootId here is be used to select one of plugin instances associated with rootID inside these SDK. https://github.com/istio/api/blob/master/extensions/v1alpha1/wasm.proto#L229-L232

Reply:
I suppose most user do not have multiple roots in one module, so we should just leave it empty here? We don't expose root id in Istio Wasm API, because we don't think people are using this.